### PR TITLE
Postgresql plugin: set plugin instance from query results.

### PR DESCRIPTION
**Pull request by rpv-tomsk**
Hi!

Proposed patches adds ability to set plugin instance from query.
This feature is useful for generic plugins.

Example:

````
<Query instance_summary>
        Statement "SELECT 'summary' as instance,\
        sum(numbackends) as numbackends, \
        sum(xact_commit) as xact_commit, \
        sum(xact_rollback) as xact_rollback, \
        sum(blks_read) as blks_read, \
        sum(blks_hit) as blks_hit, \
        sum(tup_returned) as tup_returned, \
        sum(tup_fetched) as tup_fetched, \
        sum(tup_inserted) as tup_inserted, \
        sum(tup_updated) as tup_updated, \
        sum(tup_deleted) as tup_deleted \
        FROM pg_stat_database"

        PluginInstanceFrom "instance"

        #TODO: types should have '_summary' suffix
        <Result>
                Type "pg_numbackends"
                ValuesFrom "numbackends"
        </Result>
        #
        <Result>
                Type "pg_xact"
                InstancePrefix "commit"
                ValuesFrom "xact_commit"
        </Result>
        <Result>
                Type "pg_xact"
                InstancePrefix "rollback"
                ValuesFrom "xact_rollback"
        </Result>
        #
        <Result>
                Type "pg_blks"
                InstancePrefix "blks_read"
                ValuesFrom "blks_read"
        </Result>
        <Result>
                Type "pg_blks"
                InstancePrefix "blks_hit"
                ValuesFrom "blks_hit"
        </Result>
        #
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_returned"
                ValuesFrom "tup_returned"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_fetched"
                ValuesFrom "tup_fetched"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_inserted"
                ValuesFrom "tup_inserted"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_updated"
                ValuesFrom "tup_updated"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_deleted"
                ValuesFrom "tup_deleted"
        </Result>
</Query>
<Query instance_bydb>
        Statement "SELECT datname, \
        numbackends, \
        xact_commit, xact_rollback, \
        blks_read, blks_hit, \
        tup_returned, tup_fetched, tup_inserted, tup_updated, tup_deleted \
        FROM pg_stat_database WHERE datname <> 'template0'"

        PluginInstanceFrom "datname"

        #TODO: types should have '_summary' suffix
        <Result>
                Type "pg_numbackends"
                ValuesFrom "numbackends"
        </Result>
        #
        <Result>
                Type "pg_xact"
                InstancePrefix "commit"
                ValuesFrom "xact_commit"
        </Result>
        <Result>
                Type "pg_xact"
                InstancePrefix "rollback"
                ValuesFrom "xact_rollback"
        </Result>
        #
        <Result>
                Type "pg_blks"
                InstancePrefix "blks_read"
                ValuesFrom "blks_read"
        </Result>
        <Result>
                Type "pg_blks"
                InstancePrefix "blks_hit"
                ValuesFrom "blks_hit"
        </Result>
        #
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_returned"
                ValuesFrom "tup_returned"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_fetched"
               ValuesFrom "tup_fetched"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_inserted"
                ValuesFrom "tup_inserted"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_updated"
                ValuesFrom "tup_updated"
        </Result>
        <Result>
                Type "pg_scan"
                InstancePrefix "tup_deleted"
                ValuesFrom "tup_deleted"
        </Result>
</Query>
        <Database template1>
                Host "localhost"
                Port 5432
                User "postgres"
                Password "CHANGEME"
                Query instance_summary
                Query instance_bydb
        </Database>
</Plugin>
````

### CODE REVIEW

File: src/collectd.conf.pod [hunk 1]
```diff
@@ -5192,14 +5191,42 @@ specific or global B<Interval> options).
 Please note that parameters are only supported by PostgreSQL's protocol
 version 3 and above which was introduced in version 7.4 of PostgreSQL.
 
+=item B<PluginInstanceFrom> I<column>
+
+Specify how to create the "PluginsInstance" for reporting this quiery results.
```
Comment (tokkee) [inline src/collectd.conf.pod]: Typos: PluginsInstance → PluginInstance and quiery → query

File: src/collectd.conf.pod [hunk 2]
```diff
@@ -5192,14 +5191,42 @@ specific or global B<Interval> options).
...
+Only one column is supported. You may concatenate fields and string values in
+query statement to get required results.
```
Comment (tokkee) [inline src/collectd.conf.pod]: … in *the* query statement to get *the* required results.

File: src/collectd.conf.pod [hunk 3]
```diff
@@ -5418,6 +5431,8 @@ Specify the plugin instance name that should be used instead of the database
 name (which is the default, if this option has not been specified). This
 allows to query multiple databases of the same name on the same host (e.g.
 when running multiple database server versions in parallel).
+The plugin instance name can also be set from query result using
```
Comment (tokkee) [inline src/collectd.conf.pod]: … from *the* query result using *the* PluginInstanceFrom option …

File: src/utils_db_query.c [hunk 4]
```diff
@@ -78,6 +80,7 @@ typedef struct udb_result_preparation_area_s udb_result_preparation_area_t;
 struct udb_query_preparation_area_s /* {{{ */
 {
   size_t column_num;
+  size_t plugin_instance_pos;
```
Comment (tokkee) [inline src/utils_db_query.c]: Why did you add this here instead of in udb_result_preparation_area_t along with the other position variables?

Comment (rpv-tomsk) [inline src/utils_db_query.c]: Option `PluginInstanceFrom` set on `Query` level, not inside `Result`, so  there is no need to calculate field position for each result.

**Comment by rpv-tomsk** [comment]
@tokkee, Thank you for finding typos. Updated.

File: src/utils_db_query.c (hunk 4 above)
Comment (tokkee) [inline src/utils_db_query.c]: Ah, yes, that makes perfect sense :-)

**Comment by tokkee** [comment]
:+1: LGTM

File: src/collectd.conf.pod [hunk 5]
```diff
@@ -5131,13 +5131,12 @@ The B<Query> block defines one database query which may later be used by a
...
 B<MinVersion> and B<MaxVersion> options below for an exception to this
-rule). The following configuration options are available to define the query:
+rule).
 
-In each B<Query> block, there is one or more B<Result> blocks. B<Result>
-blocks define how to handle the values returned from the query. They define
-which column holds which value and how to dispatch that value to the daemon.
-Multiple B<Result> blocks may be used to extract multiple values from a single
-query.
+In each B<Query> block, there is one or more B<Result> blocks. Multiple
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/in/are/

File: src/collectd.conf.pod [hunk 6]
```diff
@@ -5192,14 +5191,42 @@ specific or global B<Interval> options).
 Please note that parameters are only supported by PostgreSQL's protocol
 version 3 and above which was introduced in version 7.4 of PostgreSQL.
 
+=item B<PluginInstanceFrom> I<column>
+
+Specify how to create the "PluginInstance" for reporting this query results.
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/this/the/

File: src/collectd.conf.pod [hunk 7]
```diff
@@ -5192,14 +5191,42 @@ specific or global B<Interval> options).
...
+
+The I<version> has to be specified as the concatenation of the major, minor
+and patch-level versions, each represented as two-decimal-digit numbers. For
+example, version 8.2.3 will become 80203.
+
+=back
+
+The B<Result> block defines how to handle the values returned from the query.
+It defines which column holds which value and how to dispatch that value to
+the daemon.
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/the daemon/collectd/ ?

Comment (rpv-tomsk) [inline src/collectd.conf.pod]: That's just block of text what moved from line 5136.
I do not know how it would be better to write.

File: src/collectd.conf.pod (hunk 6 above)
Comment (rpv-tomsk) [inline src/collectd.conf.pod]: My branch updated, thanks.

File: src/collectd.conf.pod (hunk 5 above)
Comment (rpv-tomsk) [inline src/collectd.conf.pod]: I do not understand. s/is/are/ ???

Copy-n-pasted from line 1872:

cat collectd.conf.pod |grep -n "there is one or more"
1872:In each B<Query> block, there is one or more B<Result> blocks. B<Result> blocks
5136:In each B<Query> block, there is one or more B<Result> blocks. Multiple

Comment (rubenk) [inline src/collectd.conf.pod]: Yes, s/is/are/, sorry. "Are" is the correct grammar.

File: src/collectd.conf.pod (hunk 7 above)
Comment (rubenk) [inline src/collectd.conf.pod]: Yeah, I just think "the daemon" is a bit ambiguous, since we're really talking about collectd here.
Never mind, we can clean this up later.

File: src/collectd.conf.pod (hunk 5 above)
Comment (rpv-tomsk) [inline src/collectd.conf.pod]: Both lines updated.
